src/main/java/org/broadinstitute/hellbender/engine/AssemblyRegionIterator.java

```diff
@@ -0,0 +1,226 @@
+            throw new NoSuchElementException("next() called when there were no more elements");
+        }
+
+        previousRegion = readyRegion;
+        readyRegion = loadNextAssemblyRegion();
+        return previousRegion;
+    }
+
+    private AssemblyRegion loadNextAssemblyRegion() {
+        AssemblyRegion nextRegion = null;
+
+        while ( locusIterator.hasNext() && nextRegion == null ) {
+            final AlignmentContext pileup = locusIterator.next();
+
+            // Ignore sites that are not contained within this shard's interval
+            if ( ! readShard.getInterval().contains(pileup) ) {
+                continue;
+            }
+
+            // Pop any new pending regions off of the activity profile. These pending regions will not become ready
+            // until we've traversed all the reads that belong in them.
+            //
+            // Ordering matters here: need to check for forceConversion before adding current pileup to the activity profile
+            if ( ! activityProfile.isEmpty() ) {
+                final boolean forceConversion = pileup.getLocation().getStart() != activityProfile.getEnd() + 1;
+                pendingRegions.addAll(activityProfile.popReadyAssemblyRegions(assemblyRegionPadding, minRegionSize, maxRegionSize, forceConversion));
+            }
+
+            // Add the current pileup to the activity profile
+            if ( readShard.getPaddedInterval().contains(pileup.getLocation()) ) {
+                final SimpleInterval pileupInterval = new SimpleInterval(pileup);
+                final ReferenceContext pileupRefContext = new ReferenceContext(reference, pileupInterval);
+                final FeatureContext pileupFeatureContext = new FeatureContext(features, pileupInterval);
+
+                final ActivityProfileState profile = evaluator.isActive(pileup, pileupRefContext, pileupFeatureContext);
+                activityProfile.add(profile);
+            }
+
+            // A pending region only becomes ready once our locus iterator has advanced beyond the end of its extended span
+            // (this ensures that we've loaded all reads that belong in the new region)
+            if ( ! pendingRegions.isEmpty() && IntervalUtils.isAfter(pileup.getLocation(), pendingRegions.peek().getExtendedSpan(), readHeader.getSequenceDictionary()) ) {
+                nextRegion = pendingRegions.poll();
+            }
+        }
+
+        // When we run out of loci, close out the activity profile, and close out any remaining pending regions one at a time
+        if ( ! locusIterator.hasNext() ) {
+            if ( ! activityProfile.isEmpty() ) {
+                // Pop the activity profile a final time with forceConversion == true
+                pendingRegions.addAll(activityProfile.popReadyAssemblyRegions(assemblyRegionPadding, minRegionSize, maxRegionSize, true));
+            }
+
+            if ( ! pendingRegions.isEmpty() ) {
+                nextRegion = pendingRegions.poll();
```

Comment:
Could overwrite the nextRegion if it was specified by the first loop.

Reply:
Added in a guard